### PR TITLE
Implement DDPG

### LINKED ISSUE #3: Implement DDPG
need to refactor off-policy critic first to avoid code duplication

I will do it.

### PR DESCRIPTION
**Pull request by araffin**
## Description

Add `DDPG` as a special case of `TD3`. I took the advantage to modify SAC and TD3 so they accept any number of critics.

Missing:

- [x] Performance checks and hyperparameter tuning

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

closes #3 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have checked the codestyle using `make lint`
- [x] I have ensured `make pytest` and `make type` both pass.

### CODE REVIEW

File: tests/test_run.py [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
 
 normal_action_noise = NormalActionNoise(np.zeros(1), 0.1 * np.ones(1))
 
 
+@pytest.mark.parametrize('model_class', [TD3, DDPG])
 @pytest.mark.parametrize('action_noise', [normal_action_noise, OrnsteinUhlenbeckActionNoise(np.zeros(1), 0.1 * np.ones(1))])
-def test_td3(action_noise):
-    model = TD3('MlpPolicy', 'Pendulum-v0', policy_kwargs=dict(net_arch=[64, 64]),
-                learning_starts=100, verbose=1, create_eval_env=True, action_noise=action_noise)
+def test_dpg(model_class, action_noise):
```
Comment (AdamGleave) [inline tests/test_run.py:12]: `dpg` looks like a typo for `ddpg` but I think it's probably intended since it's both TD3 and DDPG. Can we disambiguate this? e.g. `test_deterministic_pg`?

File: stable_baselines3/ddpg/ddpg.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    :param learning_rate: (float or callable) learning rate for adam optimizer,
+        the same learning rate will be used for all networks (Q-Values, Actor and Value function)
+        it can be a function of the current progress remaining (from 1 to 0)
+    :param buffer_size: (int) size of the replay buffer
+    :param learning_starts: (int) how many steps of the model to collect transitions for before learning starts
+    :param batch_size: (int) Minibatch size for each gradient update
+    :param tau: (float) the soft update coefficient ("Polyak update", between 0 and 1)
+    :param gamma: (float) the discount factor
+    :param train_freq: (int) Update the model every ``train_freq`` steps.
+    :param gradient_steps: (int) How many gradient update after each step
```
[the remark this answers is not in the document]
Comment (araffin) [reply stable_baselines3/ddpg/ddpg.py:32]: actually, none of the two docstrings are correct ^^"
this is how many gradient steps are done after each rollout, set to -1 means you will do as many gradients as steps done in the environment

File: stable_baselines3/ddpg/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                                   n_episodes_rollout=n_episodes_rollout,
+                                   action_noise=action_noise,
+                                   policy_kwargs=policy_kwargs,
+                                   tensorboard_log=tensorboard_log,
+                                   verbose=verbose, device=device,
+                                   create_eval_env=create_eval_env, seed=seed,
+                                   optimize_memory_usage=optimize_memory_usage,
+                                   # Remove all tricks from TD3 to obtain DDPG:
+                                   # we still need to specify target_policy_noise > 0 to avoid errors
+                                   policy_delay=1, target_noise_clip=0.0, target_policy_noise=0.1,
```
[the remark this answers is not in the document]
Comment (araffin) [reply stable_baselines3/ddpg/ddpg.py:89]: just a strictly positive value, could be anything because `target_noise_clip=0.0`

[the remark this answers is not in the document]
Comment (araffin) [reply stable_baselines3/ddpg/ddpg.py:89]: 0.1 is usually the default value for std of the action noise, other than that, it's an arbitrary choice here

File: tests/test_run.py (hunk 1 above)
Comment (araffin) [reply tests/test_run.py:12]: good point, yes, not a typo, I will change that
```suggestion
def test_deterministic_pg(model_class, action_noise):
```

File: stable_baselines3/ddpg/ddpg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines3/ddpg/ddpg.py:89]: OK, thanks for the explanation!

File: stable_baselines3/ddpg/ddpg.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines3/ddpg/ddpg.py:32]: Hmm, it would be good to clarify this! But it seems not DDPG specific so perhaps better in a separate PR.

**Comment by araffin** [comment]
wait, i did not fix all your comments and i need to fux the error with the newest pytype version

**Comment by AdamGleave** [comment]
> wait, i did not fix all your comments and i need to fix the error with the newest pytype version

Ah, I thought the only thing you didn't change were the links (and I thought maybe you preferred the URL). The docstring issue with DDPG seems to be an issue with other algos so am happy to see a fix in this PR but also fine with punting it to another issue.

**Comment by AdamGleave** [comment]
I'm confused why `pytype` passes on GitLab but not on GitHub...

**Comment by araffin** [comment]
different versions...
gitlab is using a fixed docker image
